narrow: Mark messages as read for -is:dm filter

Mang0codes opened the pull request:
<!-- Describe your pull request here.-->
Messages will be marked as read when applying the -is:dm filter alone or in combination with stream and topic filters. However, they will not be marked as read if a search term is also present.

Fixes: #25113  <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! How does this PR relate to the other open PRs for this issue?

To prepare this PR for review, please work through the [guidelines on continuing unfinished work](https://zulip.readthedocs.io/en/latest/contributing/continuing-unfinished-work.html).

alya commented:
Also, please clean up your commit history. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

Mang0codes commented:
> Thanks! How does this PR relate to the other open PRs for this issue?
> 
> To prepare this PR for review, please work through the [guidelines on continuing unfinished work](https://zulip.readthedocs.io/en/latest/contributing/continuing-unfinished-work.html).

This PR is a fresh implementation for issue #25113

alya commented:
Ok, in that case please follow the instructions in the guide to explain why.

timabbott commented on web/src/filter.ts:
```diff
@@ -1128,6 +1128,24 @@ export class Filter {
+
+        if (_.isEqual(term_types, ["not-is-dm"])) {
+            return true;
+        }
```
It seems like at least this type of rule should be in `single_term_type_returns_all_messages_of_conversation`?

timabbott commented on web/src/filter.ts:
```diff
@@ -1128,6 +1128,24 @@ export class Filter {
+
+        if (_.isEqual(term_types, ["channel", "topic", "not-is-dm"])) {
+            return true;
+        }
```
I think rather than adding these here, you'll want to follow the comment at the top of this section and replace `single_term_type_returns_all_messages_of_conversation` with a function that checks that for every operator.

Mang0codes replied:
@PieterCK, I addressed this by replacing `single_term_type_returns_all_messages_of_conversation` with a function(`all_term_type_returns_all_messages_of_conversation`) that checks all operators, as per the comment.

timabbott replied:
Why not just call it `returns_all_messages_of_conversation`?

Mang0codes commented:
I wanted to make my own changes from scratch. If there’s a preferred approach, I’m open to feedback

alya commented:
As noted in the doc I linked (emphasis added):

> If you started from scratch, explain why you decided to do so, **and how your approach differs from prior work**. 

I've added some more examples and explanation to that contributor documentation [section](https://zulip.readthedocs.io/en/latest/contributing/continuing-unfinished-work.html), so please check it out. Addressing this is the next step for you getting your PR reviewed.

Mang0codes commented:
I decided to start from scratch because the previous approach was too rigid. It relied on hardcoded comparisons for specific term combinations (like ["not-is-dm"] or ["channel", "not-is-dm"]), which could easily break if new filtering conditions were added.
Instead, I introduced all_term_type_returns_all_messages_of_conversation(), which dynamically checks whether all term types fall within a valid set. As the reviewer also pointed out that checking just a single operand (has_negated_operand("is", "dm")) might not be enough. Rather than making small tweaks to single_term_type_returns_all_messages_of_conversation(), I reworked the logic to ensure it checks for different filtering cases. @alya

alya commented:
Thanks! Should be ready for review once tests pass.

Mang0codes commented:
The tests have all passed, and it's ready to be reviewed.

alya commented:
Thanks! @kennethnrk @PieterCK Would one of you have time to take a look at this one?

PieterCK commented:
I can review this today

PieterCK reviewed (commented):
@Mang0codes Thanks for picking this up! I think this is going in the right direction and the main behaviour we want is there on manual testing. I still have some questions on some changes and I think the commit description and arrangements could be better. Let me know if any of my comments are unclear

Mang0codes commented on web/src/filter.ts:
```diff
@@ -1138,55 +1138,33 @@ export class Filter {
+        ]);
 
-        return false;
+        return validCombinations.has(term_types.join(","));
```
Got it! I'll split the refactoring into a separate commit and add the new valid combinations in a follow-up commit. Thanks for the clarification!

timabbott replied:
I don't see why you need a list of combinations -- you should be able to loop through each operator and check it individually.

Also don't use `camelCase`.

Mang0codes commented on web/src/message_view_header.ts:
```diff
@@ -88,6 +88,28 @@ function get_message_view_header_context(filter: Filter | undefined): MessageVie
+            link: "/help/search-for-messages",
+        };
+    }
+
```
Okay got it. For now, I'd prefer to leave this as is and open a separate issue for it later. I'll make sure to include proper discussion and screenshots when revisiting this. Thanks for the feedback!

Mang0codes commented:
@PieterCK does this work for you? I've made all the changes you suggested.

PieterCK commented:
@Mang0codes Just checking in, are you still working on this?

timabbott commented:
Oops, I pushed the wrong branch over this; @Mang0codes would you be up for re-pushing your work? I'm not sure if @PieterCK has a local copy yet.

PieterCK commented:
yep I got a local copy of this yesterday

timabbott commented:
Closing in favor of #33642 for this issue. Thanks @Mang0codes!
